Gerrit fetcher in prow

krzyzacy opened the pull request:
This uses a gerrit-go client from github.com/andygrunwald/go-gerrit
Basically it will keep querying against target gerrit instance for new gerrit changes, and create presubmit jobs according to the prow config.

TODO:
[x] - Pass in repo it need to query against
[x] - Handle duplicated jobs? (added a storage path to store last polling timestamp)
[x] - put auth script into third party
[?] - consolidate the correct refs
[x] - add deployment file/image files that point to our gerrit instance (deployment file is going to be internal)

/assign @cjwagner

krzyzacy commented:
/lint

krzyzacy commented:
ref https://github.com/kubernetes/test-infra/issues/6840

krzyzacy commented:
```
W0215 02:12:51.580] ERROR: error loading package '': Encountered error while reading extension file 'lib.bzl': no such package '@bazel_skylib//': Error cloning repository: https://github.com/bazelbuild/bazel-skylib.git: cannot open git-upload-pack caused by https://github.com/bazelbuild/bazel-skylib.git: cannot open git-upload-pack caused by github.com
```
this looks...

krzyzacy commented:
@ixdy seems now update-deps decides to rename more BUILD -> BUILD.bazel -> https://github.com/kubernetes/test-infra/pull/6835/commits/a0fa7b6bbcff45de837ec5675a8713d4d043b15d

ixdy commented:
@krzyzacy I think that's due to #6629; that PR was created before we'd renamed everything to `BUILD.bazel`, and since the check excludes `vendor/`, they weren't renamed before merge.

krzyzacy commented:
ah ok, so I should keep the changes?

ixdy commented:
yeah, everything should be `BUILD.bazel`, unless the upstream dependency already has `BUILD` files (such as kubernetes).

BenTheElder commented on prow/cmd/gerrit/main.go:
```diff
@@ -0,0 +1,95 @@
+
+	c, err := gerrit.NewController(*instance, projs, kc, logger, ca)
+
+	tick := time.Tick(30 * time.Second)
```
consider making this configurable with a default?

BenTheElder commented on prow/cmd/gerrit/main.go:
```diff
@@ -0,0 +1,95 @@
+	}
+
+	// authenticate to gerrit server
+	if err := exec.Command("git", "clone", "https://gerrit.googlesource.com/gcompute-tools").Run(); err != nil {
```
hmm, seems like we can do this without a git clone at runtime, why not require this tool to be present and include it in the image? this smells a bit.

krzyzacy replied:
yeah, I guess I'm not sure how to do this with the prow bazel rules, or maybe I can toss up a Dockerfile with a runner?

BenTheElder replied:
with bazel you could:
- seperately publish a base image with this and reference it as your base layer
- make this tool a WORKSPACE dep or check it in and then just copy it into the bazel built image

BenTheElder replied:
it looks like these are two infrequently modified python scripts, I would just vendor the repo and copy it into the image when building

krzyzacy replied:
ack, so I need to manually `dep add` and `dep ensure` or something? (lost on dep again, probably talk to you tomorrow)

BenTheElder replied:
not sure if dep can handle go code? but if not we can put it in a third_party dir or something.

ixdy replied:
yeah, `dep` and `vendor/` are only for Go.

a `third_party` directory would work, but adding a new `git_repository` to the bazel `WORKSPACE` might be even better.

BenTheElder replied:
the downside to a `git_repository` is needing another checkout between every clean, I think we probably don't want/need that for ~300 lines of python and a license/readme

cjwagner commented on prow/cmd/gerrit/main.go:
```diff
@@ -0,0 +1,96 @@
+		select {
+		case <-tick:
+			start := time.Now()
+			if err := c.Sync(*rateLimit); err != nil {
```
nit: Like the sync period, the rate limit could be moved to the config if tweaking the ratelimit is something we will do frequently.

cjwagner commented on prow/gerrit/gerrit.go:
```diff
@@ -0,0 +1,189 @@
+
+	for _, change := range changes {
+		if err := c.ProcessChange(change); err != nil {
+			c.log.WithError(err).Error("Failed process change %v", change.CurrentRevision)
```
s/Error/Errorf/

cjwagner commented on prow/gerrit/gerrit.go:
```diff
@@ -0,0 +1,189 @@
+			},
+		}
+		pj := pjutil.NewProwJob(pjutil.PresubmitSpec(spec, kr), map[string]string{})
+		c.log.WithFields(pjutil.ProwJobFields(&pj)).Info("Creating a new prowjob.")
```
It would probably be helpful to include the change ID as a field in this log message and any other log messages that are specific to a change. That will allow us to query for the logs for actions and errors associated with a specific change.

cjwagner commented on prow/gerrit/gerrit.go:
```diff
@@ -0,0 +1,189 @@
+		return fmt.Errorf("cannot find current revision for change %v", change.ID)
+	}
+
+	for _, spec := range c.ca.Config().Presubmits[change.Project] {
```
Based on the refs below I think you want to use `fmt.Sprintf("%s/%s", c.instance, change.Project)` as the key to the presubmits map.

krzyzacy replied:
hummm, we are currently limiting one gerrit fetcher <---> one gerrit instance

the instance is going to be something like `android-review.googlesource.com` which might be too long? I guess I can trim the 1st part.

cjwagner replied:
This still needs to be addressed I think.

krzyzacy replied:
done

cjwagner commented on prow/gerrit/gerrit.go:
```diff
@@ -0,0 +1,189 @@
+// QueryChanges will query all gerrit changes since controller's last sync loop
+func (c *Controller) QueryChanges(rateLimit int) (map[string]gerrit.ChangeInfo, error) {
+	opt := &gerrit.QueryChangeOptions{}
+	opt.Query = []string{}
```
nit: This line isn't needed.

cjwagner commented on prow/gerrit/gerrit.go:
```diff
@@ -0,0 +1,189 @@
+				start++
+			} else {
+				done = true
+				break
```
Just return `pending, nil` and then you can get rid of `done`.

cjwagner commented on prow/gerrit/gerrit.go:
```diff
@@ -0,0 +1,189 @@
+		if len(*changes) == 0 {
+			break
+		}
+
```
`start` is incremented once for each change (all paths in the loop except the return path increment `start`) so you can remove the `start++` statements and put a single `start += len(changes)` here.

cjwagner commented on prow/gerrit/gerrit.go:
```diff
@@ -0,0 +1,189 @@
+}
+
+// ProcessChange creates new presubmit prowjobs base off the gerrit changes
+func (c *Controller) ProcessChange(change gerrit.ChangeInfo) error {
```
In order to handle restarts properly, this function will need to check for existing jobs and only start a new job if it hasn't been started already. (This is a TODO in the PR body.)

However, because we are querying for all changes on Gerrit after a restart, if we just list ProwJobs to look for previous runs we will end up retriggering jobs for older changes because the ProwJob records for the previous runs will have been GC'ed by sinker. I'm not sure what the best way to handle this is. We could make the Gerrit controller ignore changes older than sinker's max prowjob age, but that seems kind of gross and limiting.

krzyzacy replied:
Currently upon restart, lastUpdated will start with `time.Now()`, and should not result in duplicated triggers.

How does trigger handles this? I didn't see a similar logic in pr.go, I thought Plank would handle no trigger the already started presubmit job?

cjwagner replied:
`terminateDupes` deletes old jobs when a new one is intentionally retriggered. It isn't related to preventing a job from being retriggered improperly.
Plank doesn't suffer from this problem because it is event based instead of polling based so we won't retrigger tests for old PRs even if the prowjobs for the PR have been garbage collected because we won't rereceive events for the PR.

krzyzacy replied:
I think I can save the last updated time per sync loop, and restore from there on restart, and it should poll new changes properly

cjwagner commented on prow/gerrit/gerrit.go:
```diff
@@ -0,0 +1,189 @@
+				{
+					Number: change.Number,
+					Author: change.Owner.Name,
+					SHA:    rev.Ref,
```
I think this should be `change.CurrentRevision`.

krzyzacy replied:
I need the ref to checkout the gerrit commit, not the SHA... so I have to keep the ref here...

cjwagner commented on prow/gerrit/gerrit.go:
```diff
@@ -0,0 +1,189 @@
+			Org:     c.instance,
+			Repo:    change.Project,
+			BaseRef: rev.Ref,
+			BaseSHA: change.CurrentRevision,
```
If we use this `BaseRef` and `BaseSHA` we won't know anything about the base branch. Are we testing the feature branches without merging them with the base branch first?

krzyzacy replied:
ack, there's a `BaseChange` field I can use

ref https://godoc.org/github.com/andygrunwald/go-gerrit#ChangeInfo

cjwagner replied:
`BaseChange` is the change ID of the base change. Based on your comment above you will need the ref associated with the change ID instead of the change ID itself in order to do the checkout.
However, this will only ever test a change against the base change that it was originally created against. For the rest of our testing we only consider tests valid if they are against the most recent change to the base branch.

krzyzacy replied:
hummm, looks like there's no base ref from the API

from gerrit downloading method seems I should be able to get the PR from a single git fetch call.

krzyzacy commented on prow/cmd/gerrit/main.go:
```diff
@@ -0,0 +1,96 @@
+	}
+
+	// authenticate to gerrit server
+	if err := exec.Command("git", "clone", "https://").Run(); err != nil {
```
ack, this part is in a bad state now, I'll need to place the repo somewhere else

krzyzacy commented:
thanks for the reviews! will patch up tomorrow :-)

krzyzacy commented on prow/cmd/gerrit/main.go:
```diff
@@ -0,0 +1,96 @@
+	instance  = flag.String("gerrit-instance", "", "URL to gerrit instance")
+	projects  = flag.String("gerrit-projects", "", "Comma separated gerrit projects to fetch from the gerrit instance")
+	rateLimit = flag.Int("rate-limit", 5, "How many gerrit changes to fetch per API call")
+	interval  = flag.Duration("sync-interval", 30*time.Second, "How often to sync gerrit changes")
```
done

krzyzacy commented on prow/cmd/gerrit/main.go:
```diff
@@ -0,0 +1,96 @@
+func main() {
+	flag.Parse()
+	logrus.SetFormatter(&logrus.JSONFormatter{})
+	logger := logrus.WithField("component", "gerrit")
```
done

BenTheElder commented:
/test pull-test-infra-bazel

BenTheElder commented:
/test pull-test-infra-bazel

krzyzacy commented:
addressed most of the comments
updated:
1. moved `git-cookie-authdaemon` script into thirdparty
2. added a storage so that the fetcher can restore from last polling time

still not sure what to use for `BaseRef` here yet, maybe it can be a straight forward `git checkout` call?

ixdy commented on prow/cmd/gerrit/main.go:
```diff
@@ -0,0 +1,98 @@
+	}
+
+	// authenticate to gerrit server
+	if err := exec.Command("./git-cookie-authdaemon").Run(); err != nil {
```
does this need to be added to a docker image?

krzyzacy replied:
yeah, I haven't written the image rule yet

kargakis commented on prow/cmd/gerrit/main.go:
```diff
@@ -0,0 +1,98 @@
+limitations under the License.
+*/
+
+// package main is a gerrit fetcher updating against a target gerrit instance
```
fun fact: godoc does not work for binaries.

krzyzacy replied:
nuked

kargakis commented on prow/cmd/gerrit/main.go:
```diff
@@ -0,0 +1,98 @@
+)
+
+var (
+	configPath = flag.String("config-path", "/etc/config/config", "Path to config.yaml.")
```
Refactor this based on @stevekuznetsov's latest refactorings wrt the `options` pattern?

krzyzacy replied:
hummm, can you attach a link?

krzyzacy replied:
done I think

kargakis commented on prow/config/config.go:
```diff
@@ -128,6 +129,14 @@ type Plank struct {
+// Gerrit is config for the gerrit controller.
+type Gerrit struct {
+	// TickInterval is how ofen we do a sync with binded gerrit instance
+	TickInterval time.Duration `json:"tick_interval,omitempty"`
```
I think you need to make this a pointer if you want omitempty to work.

krzyzacy replied:
added `TickIntervalString`

BenTheElder commented:
> @krzyzacy: PR needs rebase.

nice @kargakis :+1:

krzyzacy commented:
@k8s-ci-robot rebase this PR

krzyzacy commented:
/retest

krzyzacy commented:
PTAL :-)

In the meantime I'll try to play with this somewhere secret

cjwagner commented on prow/cmd/gerrit/main.go:
```diff
@@ -0,0 +1,120 @@
+func gatherOptions() options {
+	o := options{}
+	flag.StringVar(&o.configPath, "config-path", "/etc/config/config", "Path to config.yaml.")
+	flag.StringVar(&o.instance, "gerrit-instance", "FOO", "URL to gerrit instance")
```
Default should be "".

krzyzacy replied:
whoops

krzyzacy replied:
done

cjwagner commented on prow/config/config.go:
```diff
@@ -147,6 +148,15 @@ type Plank struct {
+	TickIntervalString string        `json:"tick_interval,omitempty"`
+	TickInterval       time.Duration `json:"-"`
+	// RateLimit defines how many changes to query per gerrit API call
+	RateLimit int `json:"ratelimit,omitempty"`
```
Please mentions the default values in the GoDoc comments.

krzyzacy replied:
done, and added default

cjwagner commented on prow/gerrit/gerrit.go:
```diff
@@ -0,0 +1,210 @@
+
+		for _, change := range *changes {
+			// if we already processed this change, then we stop the current sync loop
+			layout := "2006-01-02 15:04:05"
```
nit: This could be `const`

krzyzacy replied:
done

cjwagner commented on prow/gerrit/gerrit.go:
```diff
@@ -0,0 +1,210 @@
+		}
+		pj := pjutil.NewProwJob(pjutil.PresubmitSpec(spec, kr), map[string]string{})
+		logger.WithFields(pjutil.ProwJobFields(&pj)).Info("Creating a new prowjob for.")
+		if _, err := c.kc.CreateProwJob(pj); err != nil {
```
Since we trigger PJs without regard for RunIfChanged, RunAfterSuccess, or AlwaysRun we should enforce that these fields are not used with a unit test.

krzyzacy replied:
IMO Plank will handle RunAfterSuccess right? I'll later on add support for `AlwaysRun` and `RunIfChanged`

krzyzacy replied:
added a TODO for AlwaysRun and RunIfChanged. In Gerrit I think everything will be AlwaysRun for now, as there's no plugin support to manually trigger tests yet.

krzyzacy commented on vendor/k8s.io/client-go/util/cert/BUILD:
```diff
@@ -13,6 +13,9 @@ go_library(
     ],
+    data = [
+        "testdata/dontUseThisKey.pem",
+    ],
```
hummm, running update-deps added this entry, which causes bazel test to fail?

BenTheElder replied:
is this still an issue? that's not great.

ixdy replied:
this is a mistake in upstream client-go. it shouldn't have ever added testdata on the go_library, only on the go_test.

krzyzacy commented on prow/gerrit/gerrit.go:
```diff
@@ -0,0 +1,212 @@
+			if updated.After(c.lastUpdate) {
+				// in case of a new change comes in during the current sync cycle
+				if _, ok := pending[change.CurrentRevision]; !ok {
+					pending[change.CurrentRevision] = change
```
done and updated tests

krzyzacy commented:
@cjwagner what's a better solution? Shall I make a CRD for that?

cjwagner commented:
No, I think using a PD is fine for storing state. I'd just prefer to avoid having to store state at all, but I'm not sure how to do that cleanly. It is easier with hook because we used event based triggering.

If we update the controller to report results back to Gerrit we might be able to store the state we need in Gerrit. We could comment on new PRs to indicate that tests are being triggered and provide relevant links and then only automatically triggering tests for PRs that don't have that comment. The controller could use the existence of the comment instead of the last sync time.

That is much more complicated than just storing the last updated time though, so using a PD is probably the right decision for now at least.

krzyzacy commented:
/hold cancel
I'll play with it, and add docs/patches as follow up
